sdk/core/src/main/java/com/klaviyo/coresdk/networking/requests/KlaviyoApiRequest.kt

```diff
@@ -0,0 +1,234 @@
+     * The important data to send the API request is retained.
+     *
+     * @return
+     */
+    fun toJson(): String = JSONObject()
+        .accumulate(PATH_JSON_KEY, urlPath)
+        .accumulate(METHOD_JSON_KEY, method.name)
+        .accumulate(TIME_JSON_KEY, time)
+        .accumulate(UUID_JSON_KEY, uuid)
+        .accumulate(HEADERS_JSON_KEY, headers)
+        .accumulate(QUERY_JSON_KEY, query)
+        .accumulate(BODY_JSON_KEY, body)
+        .toString()
+
+    /**
+     * To facilitate deduplication, we will treat UUID as a unique identifier
+     * such that even if a request is deserialized from JSON it is still
+     * "equal" to its original object, regardless of instance or subclass
+     *
+     * @param other
+     * @return
+     */
+    override fun equals(other: Any?): Boolean = when (other) {
+        is KlaviyoApiRequest -> equals(this, other)
+        else -> super.equals(other)
+    }
+
+    companion object {
+        const val HEADER_CONTENT = "Content-Type"
+        const val HEADER_ACCEPT = "Accept"
+        const val HEADER_REVISION = "Revision"
+
+        private const val PATH_JSON_KEY = "url_path"
+        private const val METHOD_JSON_KEY = "method"
+        private const val TIME_JSON_KEY = "time"
+        private const val UUID_JSON_KEY = "uuid"
+        private const val HEADERS_JSON_KEY = "headers"
+        private const val QUERY_JSON_KEY = "query"
+        private const val BODY_JSON_KEY = "body"
+
+        /**
+         * Construct a request from a JSON object
+         *
+         * Returns [KlaviyoApiRequest] no matter what subclass it was before encoding.
+         * This is functionally the same and saves us some trouble
+         *
+         * @param json
+         * @return
+         */
+        fun fromJson(json: JSONObject): KlaviyoApiRequest {
+            val urlPath = json.getString(PATH_JSON_KEY)
+            val method = when (json.getString(METHOD_JSON_KEY)) {
+                RequestMethod.POST.name -> RequestMethod.POST
+                else -> RequestMethod.GET
+            }
+            val time = json.getString(TIME_JSON_KEY)
+            val uuid = json.getString(UUID_JSON_KEY)
+
+            return KlaviyoApiRequest(urlPath, method, time, uuid).apply {
+                headers = json.getJSONObject(HEADERS_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                query = json.getJSONObject(QUERY_JSON_KEY).let {
+                    it.keys().asSequence().associateWith { k -> it.getString(k) }
+                }
+                body = json.optJSONObject(BODY_JSON_KEY)
+            }
+        }
+
+        /**
+         * Compare two requests purely by their UUIDs
+         *
+         * @param request1
+         * @param request2
+         * @return
+         */
+        fun equals(request1: KlaviyoApiRequest, request2: KlaviyoApiRequest): Boolean =
+            request1.uuid == request2.uuid
+    }
```

Comment:
should this just live inside the overriden equals?

Reply:
Yep, I made it static for a unit test... but I can just test the overridden equals directly 🤦